apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -874,5 +886,28 @@ private void EnlistTransaction(Transaction transaction, TimestampBound timestamp
 
         /// <inheritdoc />
         protected override DbProviderFactory DbProviderFactory => SpannerProviderFactory.Instance;
+
+        /// <summary>
+        /// Get Resource based routing flag from environment variable.
+        /// </summary>
+        /// <returns>resource based routing flage</returns>
+        private static bool GetResourceBasedRoutingFlag()
+        {
+            string resourceBaseRoutingFlagValue =
+                Environment.GetEnvironmentVariable("GOOGLE_CLOUD_ENABLE_RESOURCE_BASED_ROUTING")?.ToLower();
+            return !string.IsNullOrEmpty(resourceBaseRoutingFlagValue) &&
+                resourceBaseRoutingFlagValue.Equals("true");
+        }
+
+        /// <summary>
+        /// Override the default endpointUri with available endpointUri.
+        /// </summary>
+        /// <param name="endpointUri"></param>
+        private void OverrideSpannerClientEndpoints(string endpointUri)
+        {
+            var newBuilder = Builder.Clone();
+            newBuilder.Host = endpointUri;
```

Comment:
No matter where this logic sits at, I wouldn't override a Host explicitly specified by client code.

Reply:
logic has been as below:
1. host explicitly specified by client code then will return the same.
2. if not specified than check for resource based routing if its false return default.
3. in case of resource based routing it checks the cache in `InstanceHostManager` if available return the random `endpoint` if not available in cache it retrieve the `endpointUris` using `instanceAdminClient.GetInstance` cache it and returns random.
4. in case if `instanceAdminClient.GetInstance` returns empty array  it consider the default endpoint.

Reply:
always use first endpointUri when multiple `endpointUris` available on instance.